sqlglot/parser.py

```diff
@@ -4600,10 +4600,13 @@ def _parse_schema(self, this: t.Optional[exp.Expression] = None) -> t.Optional[e
         if self._match_set(self.SELECT_START_TOKENS):
             self._retreat(index)
             return this
-        args = self._parse_csv(lambda: self._parse_constraint() or self._parse_field_def())
+        args = self._parse_csv(self._parse_schema_item)
         self._match_r_paren()
         return self.expression(exp.Schema, this=this, expressions=args)
 
+    def _parse_schema_item(self) -> t.Optional[exp.Expression]:
+        return self._parse_constraint() or self._parse_field_def()
```

Comment:
I don't think we need to factor this out so it can be overridden in CH. We usually extend `CONSTRAINT_PARSERS` for cases similar to this one. Let's try to refactor this.

Reply:
I feel a little bit conflicted about this implementation. If I understand correctly, the `CONSTRAINT` parser requires first token to be of the `TokenType.CONSTRAINT` for "named" constraints. For those the supported syntax is:

`TokenType.CONSTRAINT id_var constraint_type expressions`

The `PROJECTION` definition syntax does not match those rules.

But there are "unnamed" constraints as well. And it looks like those are a fit - adding `PROJECTION` to `SCHEMA_UNNAMED_CONSTRAINTS` does help. But is it semantically correct? The projection is named (it has a name). Or am I misunderstanding the meaning of "unnamed"?

Reply:
That's right, good observations. I was leading to updating the parser so this `PROJECTION` constraint is treated as a `SCHEMA_UNNAMED_CONSTRAINT`.

> But is it semantically correct? The projection is named (it has a name). Or am I misunderstanding the meaning of "unnamed"?

IIRC the "unnamed" part refers to the fact that the constraint appears as a single item in the schema definition. E.g. you have some constraints that look like:

```
                this is a "named" constraint, i.e. it's associated to a column definition
                     v
CREATE TABLE z (n INT DEFAULT 0 NOT NULL)
```

So from my perspective doing this is totally fine.

Reply:
Chiming in with not much context but I think I share @GaliFFun's opinion, afaict the named constraints are basically parsed as `CONSTRAINT <name> <unnamed_constraints>` i.e the name refers to the "alias" that would be bound to it (iirc so in case it's violated it can be printed in the error logs and be associated with that rule).

 I think we can extend `_parse_constraint` to parse other named constraints that do not necessarily begin with `CONSTRAINT` ?

Reply:
Sorry for the back and forth, I took another look at the docs:

(1) It seems that CH supports [named constraints](https://clickhouse.com/docs/en/sql-reference/statements/create/table#constraints) as is already supported, and 
(2) From my understanding the `PROJECTION` is semantically closer to an index than an actual constraint, so I'd agree that it would be best to add it as an separate entry in `CONSTRAINT_PARSERS` as with other index parameters.


@GaliFFun The constraint parser doesn't _require_ the `CONSTRAINT` keyword per se, it looks for it first in order to pass the (optional) `CONSTRAINT <name>`; The rest of the expression comes from "unnamed constraints" that are defined in `CONSTRAINT_PARSERS`, but what changes is whether the unnamed constraint will be parsed in it's own node or wrapped in a `exp.ColumnConstraint`

Any opinions ^ ?

Reply:
Yeah let's do `ProjectionDef` and override it in CH, I don't think there's any point in being so pedantic in this case. There won't be any substantial difference in the AST if we treat it as a separate branch of `_parse_constraint` vs treating it as an "index".

Reply:
Updated the implementation